### PR TITLE
mgr/cephadm: only apply osd specs if the devices have changed

### PR DESCRIPTION
**Pull request by liewegas**
This adds tracking for (1) when the host devices change and (2) when osd specs were last applied to a host.  And then skips the apply if it makes sense.

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/inventory.py [hunk 1]
```diff
@@ -174,6 +174,9 @@ def find(self, service_name: Optional[str] = None) -> List[ServiceSpec]:
             service_name, specs))
         return specs
 
+    def get_created(self, spec: ServiceSpec) -> Optional[datetime.datetime]:
+        return self.spec_created.get(spec.service_name())
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:178]: this should always work. If we really want to verify the integrity here, what about

```suggestion
    def get_created(self, spec: ServiceSpec) -> Optional[datetime.datetime]:
        if spec.service_name() not in self.spec_created
            self.save(spec.service_name())
        return self.spec_created[spec.service_name()])
```

File: src/pybind/mgr/cephadm/inventory.py [hunk 2]
```diff
@@ -291,13 +301,30 @@ def update_host_daemons(self, host, dm):
...
+    def devices_changed(self, host: str, b: List[inventory.Device]) -> bool:
+        a = self.devices[host]
+        if len(a) != len(b):
+            return True
+        aj = {d.path: d.to_json() for d in a}
+        bj = {d.path: d.to_json() for d in b}
+        if aj != bj:
+            self.mgr.log.info("Detected new or changed devices on %s" % host)
+            return True
+        return False
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:315]: looks similar to 

https://github.com/ceph/ceph/blob/faa93b751dc13003b23370f769a8ea252972c3dc/src/python-common/ceph/deployment/inventory.py#L18-L19

Sadly this is in the somewhat unused `Devices` class.

File: src/pybind/mgr/cephadm/inventory.py [hunk 3]
```diff
@@ -609,6 +612,20 @@ def host_needs_new_etc_ceph_ceph_conf(self, host: str) -> bool:
...
+            or host not in self.last_device_change
+            or host not in self.last_device_update
+            or host not in self.osdspec_last_applied
+            or spec.service_name() not in self.osdspec_last_applied[host]
+        ):
+            return True
+        created = self.mgr.spec_store.get_created(spec)
+        if created and created > self.last_device_change[host]:
+            return True
+        return self.osdspec_last_applied[host][spec.service_name()] < self.last_device_change[host];
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/inventory.py:627]: otoh, `osdspec_last_applied` might be missing when upgrading.

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-02-04_11:24:34-rados:cephadm-wip-swagner4-testing-2021-02-03-1650-distro-basic-smithi/
https://pulpito.ceph.com/swagner-2021-02-05_09:14:24-rados:cephadm-wip-swagner4-testing-2021-02-03-1650-distro-basic-smithi/

* https://tracker.ceph.com/issues/49190

**Review by sebastian-philipp — APPROVED** [review]

**Pull request merged**